File: python-hypothesis/src/sort.py
from typing import List

def insertion_sort(xs:List):
    lst = xs
    n = len(lst)
    
    if n <= 1:
        return lst
    
    for i in range(1, n):
        elem = lst[i]
        j = i-1
        while j >= 0 and elem < lst[j]:
            lst[j+1] = lst[j]
            j -= 1
        lst[j+1] = elem
    
    return lst

    
def merge(lst1: List, lst2: List) -> List:
    result = []
    i = j = 0

    while i < len(lst1) and j < len(lst2):
        if lst1[i] < lst2[j]:
            result.append(lst1[i])
            i += 1
        else:
            result.append(lst2[j])
            j += 1

    result.extend(lst1[i:])
    result.extend(lst2[j:])
    return result

def merge_sort(xs: List) -> List:
    if len(xs) <= 1:
        return xs

    mid = len(xs) // 2
    left = merge_sort(xs[:mid])
    right = merge_sort(xs[mid:])
    return merge(left, right)
    
        

File: rust-quickcheck/src/sort.rs
#[cfg(test)]
mod tests {
    use quickcheck_macros::quickcheck;

    fn insertion_sort<T: Clone + Ord>(arr: &[T]) -> Vec<T> {
        let mut result: Vec<T> = Vec::new();
        'outer: for i in 0..arr.len() {
            let elem = arr[i].clone();
            for j in 0..i {
                if elem < result[j] {
                    result.insert(j, elem);
                    continue 'outer
                }
            }
            result.push(elem);
        }
        result
    }

    fn merge_sort<T: Clone + Ord>(arr: &[T]) -> Vec<T> {
        if arr.len() <= 1 {
            return Vec::from(arr);
        }
        let left = merge_sort(&arr[.. arr.len() / 2]);
        let right = merge_sort(&arr[arr.len() / 2 ..]);
        let mut result: Vec<T> = Vec::new();

        let (mut i, mut j) = (0, 0);
        while i < left.len() && j < right.len() {
            if left[i] < right[j] {
                result.push(left[i].clone());
                i += 1;
            }
            else {
                result.push(right[j].clone());
                j += 1;
            }
        }
        result.extend_from_slice(&left[i..]);
        result.extend_from_slice(&right[j..]);
        result
    }

    #[quickcheck]
    fn sort_algorithms_identical(list: Vec<i32>) -> bool {
        insertion_sort(&list) == merge_sort(&list)
    }
}
